mondrian/src/main/java/mondrian/rolap/RolapAggregator.java

```diff
@@ -26,461 +26,390 @@
-     * aggregators in that these require not oly the operand to create
-     * the aggregation String expression, but also, the aggregate table's
-     * fact count column expression.
-     * These aggregators are NOT singletons like the above aggregators; rather,
-     * each is different because of the fact count column expression.
-     */
-    public static abstract class BaseAggor extends RolapAggregator {
-        protected final String factCountExpr;
-
-        protected BaseAggor(final String name, final String factCountExpr) {
-            super(name, index++, false);
-            this.factCountExpr = factCountExpr;
-        }
+    public boolean supportsFastAggregates( Dialect.Datatype dataType ) {
+      switch ( dataType ) {
+        case Integer:
+        case Numeric:
+          return true;
+        default:
+          return false;
+      }
+    };
+
+    public Object aggregate( List<Object> rawData, Datatype datatype ) {
+      assert rawData.size() > 0;
+      switch ( datatype ) {
+        case Integer:
+          int minInt = Integer.MAX_VALUE;
+          for ( Object data : rawData ) {
+            if ( data != null ) {
+              minInt = Math.min( minInt, (Integer) data );
+            }
+          }
+          return minInt == Integer.MAX_VALUE ? null : minInt;
+        case Numeric:
+          double minDouble = Double.MAX_VALUE;
+          for ( Object data : rawData ) {
+            if ( data != null ) {
+              minDouble = Math.min( minDouble, ( (Number) data ).doubleValue() );
+            }
+          }
+          return minDouble == Double.MAX_VALUE ? null : minDouble;
+        default:
+          throw new MondrianException( "Aggregator " + this.name + " does not support datatype" + datatype.name() );
+      }
+    }
+  };
 
-        public Object aggregate(
-            Evaluator evaluator, TupleList members, Calc exp)
-        {
-            throw new UnsupportedOperationException();
-        }
+  public static final RolapAggregator Max = new RolapAggregator( "max", index++, false ) {
+    public Object aggregate( Evaluator evaluator, TupleList members, Calc exp ) {
+      return FunUtil.max( evaluator, members, exp );
+    }
 
-        public abstract boolean alwaysRequiresFactColumn();
+    public boolean supportsFastAggregates( Dialect.Datatype dataType ) {
+      switch ( dataType ) {
+        case Integer:
+        case Numeric:
+          return true;
+        default:
+          return false;
+      }
+    };
+
+    public Object aggregate( List<Object> rawData, Datatype datatype ) {
+      assert rawData.size() > 0;
+      switch ( datatype ) {
+        case Integer:
+          int maxInt = Integer.MIN_VALUE;
+          for ( Object data : rawData ) {
+            if ( data != null ) {
+              maxInt = Math.max( maxInt, (Integer) data );
+            }
+          }
+          return maxInt == Integer.MIN_VALUE ? null : maxInt;
```

Comment:
@dcleao Actually, it doesn't because Integer.MIN_VALUE is really the smallest negative integer whereas Double.MIN_VALUE is the smallest positive nonzero value.

I found out about this here:

https://stackoverflow.com/questions/3884793/why-is-double-min-value-in-not-negative#:~:text=in%20the%20documents%2C-,Double.,%2C%202%5E(%2D1074).&text=The%20mantissa%2C%20always%20a%20positive,sign%20should%20be%20multiplied%20by.

Reply:
For min and max, if the actually data ONLY contained the sentinel value, then the code will rollup to null.  This is such a corner case that I don't think we need to worry about.  If the actual data ONLY contained null, this will also roll up to null which is the correct behavior.  So, I think we are good here...

Reply:
Cool!

Reply:
> For min and max, if the actually data ONLY contained the sentinel value, then the code will rollup to null. This is such a corner case that I don't think we need to worry about.

As for this being a corner case not worth worrying about, I disagree. Sure, it's probably a rare condition, but it's still a bug, imo. Even if you then assign to it a low severity :-)